src/util/mp_arith.cpp

```diff
@@ -315,3 +319,253 @@ unsigned integer2unsigned(const mp_integer &n)
+{
+  ullong_t result=a.to_ulong()|b.to_ulong();
+  return result;
+}
+
+/*******************************************************************\
+
+Function: bitwise_and
+
+  Inputs:
+
+ Outputs:
+
+ Purpose: bitwise and
+          bitwise operations only make sense on native objects, hence the
+          largest object size should be the largest available c++ integer
+          size (currently long long)
+
+\*******************************************************************/
+
+mp_integer bitwise_and(const mp_integer &a, const mp_integer &b)
+{
+  ullong_t result=a.to_ulong()&b.to_ulong();
+  return result;
+}
+
+/*******************************************************************\
+
+Function: bitwise_xor
+
+  Inputs:
+
+ Outputs:
+
+ Purpose: bitwise xor
+          bitwise operations only make sense on native objects, hence the
+          largest object size should be the largest available c++ integer
+          size (currently long long)
+
+\*******************************************************************/
+
+mp_integer bitwise_xor(const mp_integer &a, const mp_integer &b)
+{
+  ullong_t result=a.to_ulong()^b.to_ulong();
+  return result;
+}
+
+/*******************************************************************\
+
+Function: bitwise_neg
+
+  Inputs:
+
+ Outputs:
+
+ Purpose: bitwise negation
+          bitwise operations only make sense on native objects, hence the
+          largest object size should be the largest available c++ integer
+          size (currently long long)
+
+\*******************************************************************/
+
+mp_integer bitwise_neg(const mp_integer &a)
+{
+  ullong_t result=~a.to_ulong();
+  return result;
+}
+
+/*******************************************************************\
+
+Function: arith_right_shift
+
+  Inputs:
+
+ Outputs:
+
+ Purpose: arithmetic right shift (loads sign on MSB)
+          bitwise operations only make sense on native objects, hence the
+          largest object size should be the largest available c++ integer
+          size (currently long long)
+
+\*******************************************************************/
+
+mp_integer arith_right_shift(
+  const mp_integer &a,
+  const mp_integer &b,
+  std::size_t true_size)
+{
+  llong_t number=a.to_long();
+  ullong_t shift=b.to_ulong();
+  if(shift>true_size)
+    throw "shift value out of range";
+
+  llong_t sign=(1L<<(true_size-1))&number;
+  llong_t pad=(sign==0) ? 0 : ~((1L<<(true_size-shift))-1);
+  llong_t result=(number >> shift)|pad;
+  return result;
+}
+
+/*******************************************************************\
+
+Function: left_shift
+
+  Inputs:
+
+ Outputs:
+
+ Purpose: left shift
+          bitwise operations only make sense on native objects, hence the
+          largest object size should be the largest available c++ integer
+          size (currently long long)
+
+\*******************************************************************/
+
+mp_integer left_shift(
```

Comment:
I'm not sure I fully understand this one: the newly proposed `left_shift` vs `operator<<` have some overlap in what they are trying to achieve. When would one use the former, when the latter?